pre-docs/writing_plans.md

```diff
@@ -0,0 +1,395 @@
+```
+
+
+## Plan execution functions
+Your plan can execute multiple functions on remote systems.
+
+Your plan can include functions to run commands, scripts, tasks, and other
+plans on remote nodes. These execution functions correspond to task runner
+commands.
+
+- `run_command`: Runs a command on one or more nodes.
+- `run_script`: Runs a script (a non-task executable) on one or more nodes.
+- `run_task`: Runs a task on one or more nodes.
+- `run_plan`: Runs a plan on one or more nodes.
+- `file_upload`: Uploads a file to one or more nodes.
+
+
+### Calling basic plan functions
+
+Add basic functions to your plan by calling the function with its data type and
+arguments.
+
+Basic functions in plans share a similar structure. Call these functions with
+their data type and parameters.
+
+`run_script`
+Runs a script on one or more nodes.
+`$fileref`, `$nodes`, `**$options`
+
+`file_upload`
+Uploads a file to a specified location on one or more nodes.
+`$source`, `$destination`, `$nodes`, `$options`
+
+`run_command`
+Runs a command on one or more nodes.
+`$cmd`, `$nodes`, `$options`
+
+`get_targets`
+Parses common ways of referring to targets and returns an Array of Target objects.
+`TargetSpec $targetspec`
+
+For the functions `run_script` and `file_upload`, the `$fileref` and `$source`
+parameters accept either an absolute path or a module relative path, `<MODULE
+NAME>/<FILE>` reference, which will search for `<FILE>` relative to a module's
+files directory. For example, the reference `mysql/mysqltuner.pl` searches for
+the file `<MODULES DIRECTORY>/mysql/files/mysqltuner.pl`.
+
+Note that all the `$nodes` arguments support the patterns supported by `--nodes`
+(except for shell expansion).
+
+For example, to have your plan run the script located in
+`./mymodule/files/my_script.sh` on a set of nodes, as follows. Note that these
+functions will raise an exception and stop further execution of the plan if
+they fail on any node. To prevent this and handle the error, pass the
+`_catch_errrors => true` option to the command.
+
+```
+run_script("mymodule/my_script.sh", $nodes, 'catch_errors'=> true)
+```
+
+### Target objects
+
+The Target object represents a node and its specific connection options. You
+can get a printable representation via the name function, as well as access
+components of the target: `protocol`, `host`, `port`, `user`, `password`. You can also
+assign variables (`set_var`) to a target and get a list of existing variables the
+target has (`target.vars`).
+
+Set vars in a plan using `$target.set_var`.
+
+
+```
+plan vars(String $host) {
+	$target = get_targets($host)[0]
+	$target.set_var('operatingsystem', 'windows')
+	$targetvars = $target.vars
+	run_command("echo 'Vars for ${host}: ${$targetvars}'", $host)
+}
+```
+
+Or set `vars` in the inventory file using the `vars` key at the group level.
+
+```yaml
+groups:
+  - name: my_nodes
+	nodes:
```

Comment:
Maybe remove whitespace?

Reply:
Be nice to fix this. I don't see any other issues.

Reply:
This indentation is messed up on the live site too(though in a different way).
It will need to be fixed in a separate PR so it gets back into easyDITA